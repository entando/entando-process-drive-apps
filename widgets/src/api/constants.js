export const DOMAINS = {
  PDA: process.env.REACT_APP_DOMAIN || '',
  APP_BUILDER: process.env.REACT_APP_APP_BUILDER_DOMAIN || '',
};

export const IS_MOCKED_API = process.env.REACT_APP_MOCKED_API === 'true';

export const MOCK_API_DELAY = 800;

export const LOCAL = process.env.REACT_APP_LOCAL === 'true';

export const METHODS = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE',
};

// Inject ur access_token if you are trying to reach the APP-BUILDER API locally
export const TOKENS = {
  APP_BUILDER: '0694821a5bdc2229aac820e8a773669a',
};
