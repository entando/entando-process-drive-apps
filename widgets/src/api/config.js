import { DOMAIN, IS_MOCKED_API, MOCK_API_DELAY } from 'api/constants';
import utils from 'utils';
import config from 'mocks/config';

const get = async widgetCode => {
  if (IS_MOCKED_API) {
    await utils.timeout(MOCK_API_DELAY);
    return config;
  }

  const url = `${DOMAIN}/configs/${widgetCode}`;
  const response = await fetch(url);

  return response.json();
};

export default {
  get,
};
