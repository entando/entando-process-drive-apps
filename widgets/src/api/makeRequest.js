import { IS_MOCKED_API, MOCK_API_DELAY, DOMAINS, METHODS } from 'api/constants';
import utils from 'utils';

const getParams = queryParams =>
  Object.keys(queryParams)
    .filter(k => queryParams[k] !== undefined && queryParams[k] !== '')
    .map(k => `${encodeURIComponent(k)}=${encodeURIComponent(queryParams[k])}`)
    .join('&');

export default async ({
  domain,
  uri,
  method,
  mockResponse = {},
  useAuthentication,
  body,
  headers = {},
  queryParams,
  forceMock,
}) => {
  if (IS_MOCKED_API || forceMock) {
    await utils.timeout(MOCK_API_DELAY);
    return mockResponse;
  }

  const url = `${domain || DOMAINS.PDA}${uri}${queryParams ? `?${getParams(queryParams)}` : ''}`;

  const requestHeaders = useAuthentication
    ? new Headers({
        'Content-Type': 'application/json',
        Authorization: `Bearer ${localStorage.getItem('token')}`,
        ...headers,
      })
    : new Headers(headers);

  const configs = {
    method: method || METHODS.GET,
    ...(body ? { body } : {}),
    headers: requestHeaders,
  };

  const response = await fetch(url, configs);
  if (!response.ok) {
    const payload = await response.json();
    throw new Error(
      payload.message
        ? payload.message
        : `Bad Response from server: ${response.status} ${response.statusText}`
    );
  }

  const responseHeaders = response.headers.get('Content-Type');

  if (responseHeaders && responseHeaders.includes('xml')) {
    return response.text();
  }
  if (responseHeaders && responseHeaders.includes('octet')) {
    return response;
  }
  return response.json();
};
