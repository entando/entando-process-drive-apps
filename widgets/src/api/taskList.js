import { IS_MOCKED_API, MOCK_API_DELAY, SERVICE } from 'api/constants';
import getTasks from 'mocks/taskList/api.mock';
import utils from 'utils';

const get = async (connection, page, pageSize, sortedColumn, sortOrder, filter) => {
  if (IS_MOCKED_API) {
    await utils.timeout(MOCK_API_DELAY);
    return getTasks(page, pageSize, sortedColumn, sortOrder, filter);
  }

  const url = `${SERVICE.URL}/connections/${connection}/tasks?sort=taskId`;
  const token = localStorage.getItem('token');
  const response = await fetch(url, {
    method: 'get',
    headers: new Headers({
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    }),
  });

  return response.json();
};

export default {
  get,
};
