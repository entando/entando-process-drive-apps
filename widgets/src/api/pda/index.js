import makeRequest from 'api/makeRequest';
import { SERVICE, METHODS } from 'api/constants';

import MOCK_CONNECTIONS from 'mocks/pda/connections';
import MOCK_PROCESSES from 'mocks/pda/processes';

export const getConnections = async () =>
  makeRequest({
    domain: SERVICE.URL,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: MOCK_CONNECTIONS,
    useAuthentication: false,
  });

export const getProcess = async connection =>
  makeRequest({
    domain: SERVICE.URL,
    uri: `/connections/${connection}/processes/definitions`,
    method: METHODS.GET,
    mockResponse: MOCK_PROCESSES,
    useAuthentication: false,
  });
