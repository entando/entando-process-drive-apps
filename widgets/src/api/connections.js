import { METHODS, DOMAINS } from 'api/constants';

import { CONNECTIONS } from 'mocks/taskList/configs';
import makeRequest from 'api/makeRequest';

// eslint-disable-next-line import/prefer-default-export
export const getConnections = async () =>
  makeRequest({
    domain: DOMAINS.PDA,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: CONNECTIONS,
    useAuthentication: false,
  });
