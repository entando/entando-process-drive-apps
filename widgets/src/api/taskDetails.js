import { SERVICE, METHODS } from 'api/constants';
import mockedTask from 'mocks/taskDetails/task';

import { CONNECTIONS, PROCESS, GROUPS, COLUMNS } from 'mocks/taskList/configs';
import makeRequest from 'api/makeRequest';

export const getTask = async (connection, taskId) => {
  return makeRequest({
    domain: SERVICE.URL,
    uri: `/connections/${connection}/tasks/${taskId}`,
    method: METHODS.GET,
    mockResponse: mockedTask,
    useAuthentication: true,
  });
};

// Configs

export const getConnections = async () =>
  makeRequest({
    domain: SERVICE.URL,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: CONNECTIONS,
    useAuthentication: false,
  });

export const getProcess = async connection =>
  makeRequest({
    // domain: PDA_DOMAIN,
    uri: `/connections/${connection}/processes/definitions`,
    method: METHODS.GET,
    mockResponse: PROCESS,
    useAuthentication: true,
  });

export const getGroups = async connection =>
  makeRequest({
    // domain: PDA_DOMAIN,
    uri: `/connections/${connection}/groups`,
    method: METHODS.GET,
    mockResponse: GROUPS,
    useAuthentication: true,
  });

export const getColumns = async connection =>
  makeRequest({
    // domain: PDA_DOMAIN,
    uri: `/connections/${connection}/tasks/columns`,
    method: METHODS.GET,
    mockResponse: COLUMNS,
    useAuthentication: true,
  });
