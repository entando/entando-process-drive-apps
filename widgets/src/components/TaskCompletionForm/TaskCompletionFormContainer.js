import React from 'react';
import PropTypes from 'prop-types';
import i18next from 'i18next';
import { ThemeProvider } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';

import theme from 'theme';
import withAuth from 'components/common/auth/withAuth';
import { getTask, getTaskForm, postTaskForm } from 'api/pda/tasks';
import { getPageWidget } from 'api/app-builder/pages';
import CustomEventContext from 'components/common/CustomEventContext';
import WidgetBox from 'components/common/WidgetBox';
import JSONForm from 'components/common/form/JSONForm';
import utils from '../../utils';

class TaskCompletionFormContainer extends React.Component {
  static extractProperties(node, nodeName = '', path = '') {
    // if node.type is object - it has own properties
    if (!node) {
      return [];
    }

    if (node.type === 'object') {
      const nodesProperties = Object.keys(node.properties);
      const paths = nodesProperties.map(property =>
        TaskCompletionFormContainer.extractProperties(
          node.properties[property],
          property,
          path ? `${path}.${nodeName}` : nodeName
        )
      );
      return paths.flat();
    }

    // if node.type is not an object - it a final property
    return path ? `${path}.${nodeName}` : nodeName;
  }

  constructor(props) {
    super(props);

    this.state = {
      config: null,
      loading: false,
      submitting: false,
      formSchema: null,
      formData: {},
    };

    this.fetchForm = this.fetchForm.bind(this);
    this.fetchTaskFormData = this.fetchTaskFormData.bind(this);
    this.fetchSchema = this.fetchSchema.bind(this);
    this.submitProcessForm = this.submitProcessForm.bind(this);
  }

  componentDidMount() {
    this.setState({ loading: true }, async () => {
      const config = await this.fetchWidgetConfigs();
      this.setState({ config }, this.fetchForm);
    });
  }

  componentDidUpdate = prevProps => {
    const { taskId } = this.props;
    if (prevProps.taskId !== taskId) {
      this.fetchForm();
    }
  };

  async fetchWidgetConfigs() {
    const { pageCode, frameId } = this.props;
    try {
      // config will be fetched from app-builder
      const widgetConfigs = await getPageWidget(pageCode, frameId);
      if (widgetConfigs.errors && widgetConfigs.errors.length) {
        throw widgetConfigs.errors[0];
      }
      const { config } = widgetConfigs.payload;
      const settings = (config.settings && JSON.parse(config.settings)) || {};

      const parsedSettings = Object.keys(settings).reduce(
        (acc, settingKey) => ({
          ...acc,
          [settingKey]: JSON.parse(settings[settingKey]),
        }),
        {}
      );

      return {
        ...config,
        settings: parsedSettings,
      };
    } catch (error) {
      this.handleError(error.message);
    }
    return null;
  }

  async fetchForm() {
    this.setState({ loading: true });
    const formDataPromise = this.fetchTaskFormData();
    const formSchemaPromise = this.fetchSchema();

    const taskData = await formDataPromise;
    const formSchema = await formSchemaPromise;

    const properties = TaskCompletionFormContainer.extractProperties(formSchema);

    const mergedData = { ...taskData.inputData, ...taskData.outputData };

    const formData = Object.keys(mergedData).reduce((acc, property) => {
      if (properties.includes(property)) {
        return { ...acc, [property]: mergedData[property] };
      }
      return acc;
    }, {});

    // TEMP PAM BUG FIX, REMOVE THIS WHEN FIXED ON PAM, PASS formData to this.setState
    const modifiedFormData = Object.keys(formData).reduce((acc, field) => {
      const path = field.split('.');

      let property = formSchema;
      for (let i = 0; i < path.length; i += 1) {
        property = property.properties[path[i]];
      }

      if (property.type === 'array') {
        if (!Array.isArray(formData[field])) {
          return { ...acc, [field]: [formData[field]] };
        }
      }

      return { ...acc, [field]: formData[field] };
    }, {});
    // ^^^ TEMP PAM BUG FIX, REMOVE THIS WHEN FIXED ON PAM, PASS formData TO this.setState()

    this.setState({ formData: utils.unflat(modifiedFormData), formSchema, loading: false });
  }

  async fetchTaskFormData() {
    const { config } = this.state;
    const { taskId } = this.props;

    const connection = config && config.knowledgeSource;

    if (!connection) {
      this.handleError(i18next.t('messages.errors.noConnection'));
    } else {
      try {
        const task = await getTask(connection, taskId);

        return (
          (task &&
            task.payload && {
              outputData: task.payload.outputData,
              inputData: task.payload.inputData,
            }) || { outputData: {}, inputData: {} }
        );
      } catch (error) {
        this.handleError(error.message);
      }
    }
    return {};
  }

  async fetchSchema() {
    const { config } = this.state;
    const { taskId } = this.props;
    const connection = (config && config.knowledgeSource) || '';

    try {
      const formSchema = await getTaskForm(connection, taskId);
      return formSchema;
    } catch (error) {
      this.handleError(error.message);
    }
    return null;
  }

  submitProcessForm(form) {
    this.setState({ submitting: true }, async () => {
      const { config } = this.state;
      const { taskId, onSubmitForm } = this.props;
      const connection = (config && config.knowledgeSource) || '';

      try {
        const response = await postTaskForm(connection, taskId, form.formData);
        onSubmitForm({ ...form, response });
      } catch (error) {
        this.handleError(error.message);
      } finally {
        this.setState({ submitting: false });
      }
    });
  }

  handleError(err) {
    const { onError } = this.props;
    onError(err);
  }

  render() {
    const { loading, submitting, formData, formSchema, config } = this.state;
    const { onError } = this.props;

    const uiSchemas = (config && config.settings && config.settings.uiSchemas) || [];
    const defaultColumnSize =
      (config && config.settings && config.settings.defaultColumnSize) || 12;

    return (
      <CustomEventContext.Provider value={{ onSubmitForm: this.submitProcessForm, onError }}>
        <ThemeProvider theme={theme}>
          <Container maxWidth={false} disableGutters>
            <WidgetBox>
              <JSONForm
                loading={loading}
                formSchema={formSchema}
                formData={formData}
                uiSchemas={uiSchemas}
                submitting={submitting}
                defaultColumnSize={defaultColumnSize}
              />
            </WidgetBox>
          </Container>
        </ThemeProvider>
      </CustomEventContext.Provider>
    );
  }
}

TaskCompletionFormContainer.propTypes = {
  taskId: PropTypes.string.isRequired,
  onError: PropTypes.func,
  onSubmitForm: PropTypes.func,
  pageCode: PropTypes.string,
  frameId: PropTypes.string,
};

TaskCompletionFormContainer.defaultProps = {
  onError: () => {},
  onSubmitForm: () => {},
  pageCode: '',
  frameId: '',
};

export default withAuth(TaskCompletionFormContainer, [
  'task-get',
  'task-form-get',
  'task-form-submit',
]);
