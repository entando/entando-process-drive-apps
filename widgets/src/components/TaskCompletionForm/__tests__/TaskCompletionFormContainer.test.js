import { render, wait } from '@testing-library/react';
import React from 'react';
import 'mocks/i18nMock';

import { DOMAINS } from 'api/constants';
import TaskCompletionFormContainer from 'components/TaskCompletionForm/TaskCompletionFormContainer';
import WIDGET_CONFIGS from 'mocks/app-builder/pages';
import MOCKED_GET_TASK_RESPONSE from 'mocks/taskDetails/getTask';
import MOCKED_GET_TASK_FORM_RESPONSE from 'mocks/taskCompletionForm/getFormSchema';

describe('<TaskCompletionFormContainer />', () => {
  it('renders snapshot correctly', async () => {
    const configUrl = `${DOMAINS.APP_BUILDER}/api/pages//widgets/`;
    const connection = 'kieStaging';
    const taskId = 'test';
    const taskListUrl = `${DOMAINS.PDA}/connections/${connection}/tasks/${taskId}@${
      MOCKED_GET_TASK_RESPONSE.payload.id.split('@')[1]
    }`;

    fetch
      .once(JSON.stringify(WIDGET_CONFIGS.TASK_DETAILS))
      .once(JSON.stringify(MOCKED_GET_TASK_RESPONSE))
      .once(JSON.stringify(MOCKED_GET_TASK_FORM_RESPONSE));

    const { container } = render(<TaskCompletionFormContainer taskId={taskId} />);

    await wait(() => expect(container).toMatchSnapshot());

    expect(fetch.mock.calls.length).toBe(3);
    expect(fetch.mock.calls[0][0]).toEqual(configUrl);
    expect(fetch.mock.calls[1][0]).toEqual(taskListUrl);
  });

  it('renders snapshot correctly on error state', async () => {
    const { container } = render(<TaskCompletionFormContainer taskId="1" />);

    await wait(() => expect(container).toMatchSnapshot());
  });
});
