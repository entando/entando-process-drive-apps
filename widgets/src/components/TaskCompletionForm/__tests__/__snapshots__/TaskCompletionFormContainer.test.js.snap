// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskCompletionFormContainer /> renders snapshot correctly 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-disableGutters MuiContainer-maxWidthLg"
  >
    <div
      class="MuiPaper-root WidgetBox-widgetBox-9 MuiPaper-outlined"
    >
      <div
        class="MuiBox-root MuiBox-root-39"
      >
        <div>
          <form
            class="JSONForm-themedForm-42"
          >
            <div
              class="makeStyles-container-63"
            >
              <div>
                <div>
                  
                  <hr
                    class="MuiDivider-root-73 makeStyles-divider-69"
                  />
                  
                  <div
                    class="MuiGrid-root-80 makeStyles-gridContainer-68 MuiGrid-container-81 MuiGrid-spacing-xs-3-105"
                  >
                    <div
                      class="MuiGrid-root-80 makeStyles-gridItem-71 makeStyles-gridItem-72 MuiGrid-item-82 MuiGrid-grid-xs-12-126"
                    >
                      <div
                        class="makeStyles-container-63"
                      >
                        <div>
                          <div>
                            PerformanceEvaluation-taskform.frm
                            <hr
                              class="MuiDivider-root-73 makeStyles-divider-69"
                            />
                            
                            <div
                              class="MuiGrid-root-80 makeStyles-gridContainer-68 MuiGrid-container-81 MuiGrid-spacing-xs-3-105"
                            >
                              <div
                                class="MuiGrid-root-80 makeStyles-gridItem-71 makeStyles-gridItem-183 MuiGrid-item-82 MuiGrid-grid-xs-12-126"
                              >
                                <div
                                  class="makeStyles-container-63"
                                >
                                  <label
                                    class="makeStyles-label-64"
                                    for="root_47078d21-7da5-4d3f-8355-0fcd78b09f39_reason"
                                  >
                                    Reason
                                  </label>
                                  <div>
                                    <div
                                      class="MuiFormControl-root-184 MuiFormControl-fullWidth-187"
                                    >
                                      <div
                                        class="MuiFormControl-root-184 MuiTextField-root-188"
                                      >
                                        <label
                                          class="MuiFormLabel-root-201 MuiInputLabel-root-189 MuiInputLabel-formControl-195 MuiInputLabel-animated-198 MuiInputLabel-shrink-197 MuiInputLabel-outlined-200 MuiFormLabel-filled-206"
                                          data-shrink="true"
                                          for="root_47078d21-7da5-4d3f-8355-0fcd78b09f39_reason"
                                          id="root_47078d21-7da5-4d3f-8355-0fcd78b09f39_reason-label"
                                        >
                                          Reason
                                        </label>
                                        <div
                                          class="MuiInputBase-root-224 MuiOutlinedInput-root-209 MuiInputBase-formControl-225"
                                        >
                                          <input
                                            aria-invalid="false"
                                            class="MuiInputBase-input-235 MuiOutlinedInput-input-219 MuiInputBase-inputHiddenLabel-241"
                                            id="root_47078d21-7da5-4d3f-8355-0fcd78b09f39_reason"
                                            type="string"
                                            value="Showed great results!"
                                          />
                                          <fieldset
                                            aria-hidden="true"
                                            class="PrivateNotchedOutline-root-242 MuiOutlinedInput-notchedOutline-218"
                                          >
                                            <legend
                                              class="PrivateNotchedOutline-legendLabelled-244"
                                            >
                                              <span>
                                                Reason
                                              </span>
                                            </legend>
                                          </fieldset>
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                              <div
                                class="MuiGrid-root-80 makeStyles-gridItem-71 makeStyles-gridItem-183 MuiGrid-item-82 MuiGrid-grid-xs-12-126"
                              >
                                <div
                                  class="makeStyles-container-63"
                                >
                                  <label
                                    class="makeStyles-label-64"
                                    for="root_47078d21-7da5-4d3f-8355-0fcd78b09f39_performance"
                                  >
                                    Performance
                                     *
                                  </label>
                                  <div>
                                    <div
                                      class="MuiFormControl-root-184 MuiFormControl-fullWidth-187"
                                    >
                                      <div
                                        class="MuiFormControl-root-184 MuiTextField-root-188"
                                      >
                                        <label
                                          class="MuiFormLabel-root-201 MuiInputLabel-root-189 MuiInputLabel-formControl-195 MuiInputLabel-animated-198 MuiInputLabel-shrink-197 MuiInputLabel-outlined-200 MuiFormLabel-filled-206 Mui-required Mui-required"
                                          data-shrink="true"
                                          for="root_47078d21-7da5-4d3f-8355-0fcd78b09f39_performance"
                                          id="root_47078d21-7da5-4d3f-8355-0fcd78b09f39_performance-label"
                                        >
                                          Performance
                                          <span
                                            class="MuiFormLabel-asterisk-208 MuiInputLabel-asterisk-194"
                                          >
                                             
                                            *
                                          </span>
                                        </label>
                                        <div
                                          class="MuiInputBase-root-224 MuiOutlinedInput-root-209 MuiInputBase-formControl-225"
                                        >
                                          <input
                                            aria-invalid="false"
                                            class="MuiInputBase-input-235 MuiOutlinedInput-input-219 MuiInputBase-inputHiddenLabel-241"
                                            id="root_47078d21-7da5-4d3f-8355-0fcd78b09f39_performance"
                                            required=""
                                            type="integer"
                                            value="100"
                                          />
                                          <fieldset
                                            aria-hidden="true"
                                            class="PrivateNotchedOutline-root-242 MuiOutlinedInput-notchedOutline-218"
                                          >
                                            <legend
                                              class="PrivateNotchedOutline-legendLabelled-244"
                                            >
                                              <span>
                                                Performance
                                                 *
                                              </span>
                                            </legend>
                                          </fieldset>
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="JSONForm-actionButtons-40"
            >
              <button
                class="MuiButtonBase-root-275 MuiButton-root-246 MuiButton-contained-254 MuiButton-containedPrimary-255"
                tabindex="0"
                type="submit"
              >
                <span
                  class="MuiButton-label-247"
                >
                  common.submit
                </span>
                <span
                  class="MuiTouchRipple-root-278"
                />
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<TaskCompletionFormContainer /> renders snapshot correctly on error state 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-disableGutters MuiContainer-maxWidthLg"
  >
    <div
      class="MuiPaper-root WidgetBox-widgetBox-296 MuiPaper-outlined"
    >
      <div
        class="MuiBox-root MuiBox-root-326"
      >
        messages.warnings.noFormSchema
      </div>
    </div>
  </div>
</div>
`;
